src/object.cr

```diff
@@ -86,72 +86,70 @@ class Object
   # ```
-  # ary = ["one", "two", "three, etc."]
-  # ary.inspect # => ["one", "two", "three, etc."]
+  # "foo #{bar} baz" # calls bar.to_io(io)
   # ```
   #
-  # Note that if Array invoked `#to_s` on each of the elements
-  # above, the output would have been this:
-  #
+  # `IO#<<` calls this method to append an object to itself:
   # ```
-  # ary = ["one", "two", "three, etc."]
-  # # If inspect invoked to_s on each element...
-  # ary.inspect # => [one, two, three, etc.]
+  # io << bar # calls bar.to_s(io)
   # ```
   #
-  # Note that it's not clear how many elements the array has,
-  # or which are they, because `#to_s` doesn't guarantee that
-  # the string representation is clearly delimited (in the case
-  # of `String` the quotes are not shown).
+  # Thus implementations must not interpolate `self` in a string literal or call
+  # `io << self` which both would lead to an endless loop.
   #
-  # Also note that sometimes the output of `#inspect` will look
-  # like a Crystal expression that will compile, but this isn't
-  # always the case, nor is it necessary. Notably, `Reference#inspect`
-  # and `Struct#inspect` return values that don't compile.
+  # Also see `#inspect(IO)`.
+  abstract def to_s(io : IO) : Nil
+
+  # Returns an unambiguous and information-rich string representation of this
+  # object, typically intended for developers.
+  #
+  # This method should usually **not** be overridden. It delegates to
+  # `#inspect(IO)` which can be overridden for custom implementations.
   #
-  # Classes must usually **not** override this method. Instead,
-  # they must override `inspect(io)`, which must append to the
-  # given `IO` object.
+  # Also see `#to_s`.
   def inspect : String
     String.build do |io|
       inspect io
     end
   end
 
-  # Appends a string representation of this object
-  # to the given `IO` object.
+  # Prints an unambiguous and information-rich string representation of this
+  # object, typically intended for developers, to *io*.
+  #
+  # The representation should be clearly delimited.
+  # A Crystal expression that recreates an object with the same value (given an
+  # identical environment) is an ideal representation. It doesn't need to be
+  # actual code that compiles, but a resemblance is recommended.
+  # When this is not possible (for example for non-printable internal state)
+  # another useful description will do.
+  #
+  # It is advisable to have an appropriate `#inspect` implementation on every
+  # type. Default implementations are provided by `Struct#inspect` and
+  # `Reference#inspect`.
+  #
+  # It is similar to `#to_s(IO)`, but often provides more information.
+  # For types that don't provide a custom implementation of this method,
+  # default implementation delegates to `#to_s(IO)`.
   #
-  # Similar to `to_s(io)`, but usually appends more information
-  # about this object.
-  # See `#inspect`.
+  # `::p` and `::p!` use this method to present an object.
```

Comment:
Is it necessary to have the `::` here? `doc` doesn't resolve it correctly? More a curiosity than a request to change.

```suggestion
  # `p` and `p!` use this method to present an object.
```

Reply:
Yeah otherwise the docs generator wouldn't catch the reference to `p`.